File: Euler12.py
import numpy as np
from functools import reduce
from time import perf_counter

def EratMa(N):
    """
    The Sieve of Eratosthenes implemented with numpy arrays
    Returns the list of primes lower or equal than N
    """
    if N < 11:
        first_primes = ([],[],[2],[2,3],[2,3],[2,3,5],[2,3,5],[2,3,5,7],[2,3,5,7],[2,3,5,7],[2,3,5,7])
        return(first_primes[N]) 
    N_bool = np.array([True]*(N+1))
    N_bool[0] = False
    N_bool[1] = False
    N_bool[2**2::2] = False # eiminating even numbers
    N_bool[3**2::3] = False # eiminating multiples of 3
    p = 5
    while p*p <= N:
        # start from p**2
        # because all the smaller composites have factors <p
        # and are already eliminated in previous steps
        if N_bool[p]:
            N_bool[p**2::p] = False # python doesn't care if ::p goes beyond existing array
        if N_bool[p+2]:
            N_bool[(p+2)**2::(p+2)] = False
        p = p + 6 # we can move in steps of 6
    # returns indices of nonzero elements, which in this case
    # ARE the correcponding natural numbers, which were not eliminated
    # Have to use index [0], because for technical reasons it produces a 2D array
    return np.nonzero(N_bool)[0] # in this version returns np.array


def IncompletePrimeFactorExtP(N, primes0): 
    """
    Returns the list of all prime factors of N with a possible exception
    of a single factor > sqrt(N). If N is prime, it returns an empty list.
    
    In this version primes must be proveded as an external np.array
    that goes at least up to sqrt(N)
    """
    # create the list of primes lower or equal to sqrt(N)
    
    idx = np.searchsorted(primes0, int(np.floor(np.sqrt(N))), side="left") 
    # np.searchsorted(a, v, side = "left/right")
    # returns index i corresponding to:
    # left: a[i-1] < v <= a[i]
    # right: a[i-1] <= v < a[i]
    
    # it does not matter if we include a signle prime larger than sqrt(N)
    # so there is no need to fine tune to idx or idx-1
    
    primes = primes0[:idx+1]
    
    # if N is divisible by a prime, N%p = 0
    # so when we create an array of N%p for all primes
    # it has 0 entries for factors and non-zero one for non-factors
    # recasting it as bool turns 0 into False and non-zero to True
    # after we invert them, we get True for factors and False for non-factors
    prime_mask = np.invert(np.array(N%primes, dtype = bool))
    # now we only need to apply this mask to return only prime factors
    # but since we limited ourselves to p < sqrt(N)
    # we might miss a factor > sqrt(N), like e.g. 33 = 3*11
    # so be careful when using this output
    return list(primes[prime_mask])

def FactorMultiplicityExtPndiv(N, primes0):
    """
    Returns two lists, the first with all the factors of N
    the second with corresponding multimlicities of each of them.
    And also it calculates and returns the number of divisors
    
    In this version primes must be proveded as an external np.array
    that goes at least up to sqrt(N)
    """
    factors = IncompletePrimeFactorExtP(N, primes0)
    if not factors: return [N],[1],2 # N is prime
    multiplicity = []
    # just cycle through all the factors 
    # and check how many time N is divisible by each
    for p in factors:
        m = 0
        while N%p == 0:
            m += 1
            N = N//p
        multiplicity.append(m)
    if N != 1: # the case of a single additional factor > sqrt(N)
        factors.append(N)
        multiplicity.append(1)
        
    # must recast them as python int, because
    # since we used numpy array before, their
    # type was changed to long_scalars, the default
    # int type of numpy. If we don't do this
    # we quickly get overflow when multiplying these
    # numbers
    
    factors = [int(x) for x in factors]
    # number of divisors for multiplicities m1, m2, m3... is (m1+1)*(m2+1)*(m3+1)...
    ndiv = reduce(lambda x, y: x*y, [multiplicity[x]+1 for x in range(len(multiplicity))], 1)

    return factors, multiplicity, ndiv


start = perf_counter()  
limdiv = 500
primes = EratMa(1000) # only need up to sqrt(n)
n = 1
ndiv = 0
while ndiv < limdiv: # n up to 1 000 000 means T up to 500 000*1 000 001 = 500 000 500 000 ~ 5*10**11

    # T = n(n+1)/2
    # either n or n+1 is even
    # and they don't have common divisors
    
    if n % 2 == 0:
        ndiv = FactorMultiplicityExtPndiv(n+1, primes)[2] * FactorMultiplicityExtPndiv(n//2, primes)[2]
        # there are no common prime factors in numbers n and n+1 
        # (can be proved by the sieve of Eratosthenes logic)
        # so the multiplicities are independent
        # and the numbers of divisors are simply multiplied by each other
        # to get the number of divisors for the composite
    else:
        ndiv = FactorMultiplicityExtPndiv(n, primes)[2] * FactorMultiplicityExtPndiv((n + 1)//2, primes)[2]
    
    n += 1;
    
end = perf_counter()

print("n =", n, "T =", n*(n+1)//2, "number of divisors =", ndiv)
print(end - start,'sec')

File: Euler12.cpp
#include <cmath>
#include <ctime>
#include <vector>
#include <iostream>
#include <algorithm>

using namespace std;


//the sieve of Eratosthenes, the limit by both memory and time is about 10^10
// older version with step 2
// (10^10 takes about 137 seconds)
vector<unsigned long> Erat(unsigned long N){
    vector<unsigned long> primes;
    if (N == 0 || N == 1) return primes;
    primes.push_back(2);
    if (N == 2) return primes;
    primes.push_back(3);
    if (N == 3) return primes;
    vector<bool> Nbool(N+1, true);
    vector<bool>::iterator itr = Nbool.begin();
    *itr = false;
    *(itr+1) = false;
    for (itr = Nbool.begin() + 4; itr < Nbool.end(); itr += 2) *itr = false;
    unsigned long p = 3;
    while (p*p <= N) {
        for (itr = Nbool.begin() + p*p; itr < Nbool.end(); itr += p) *itr = false;
        p += 2;
        while (! Nbool[p]) p += 2;
    }
    for (itr = Nbool.begin()+5; itr < Nbool.end(); itr += 2) {
        if (*itr) primes.push_back(itr - Nbool.begin());
    }
    return primes;
}

vector<unsigned long> IncompletePrimeFactor(unsigned long N, vector<unsigned long> primes0) 
// this one with externally supplied primes, need them to be at least up to sqrt(N)
{
    vector<unsigned long>::iterator itr, idx;
    vector<unsigned long> factors;
    idx = lower_bound(primes0.begin(), primes0.end(), (unsigned long)(floor(sqrt((double) N))));
    // determine the itr below the value for sqrt(N)
    for (itr = primes0.begin(); itr != idx + 1; itr ++) // add +1 in case sqrt(N) is a prime
    {
        if (N % *itr == 0) factors.push_back(*itr);
    }
    return factors;
} 

unsigned long FactorMultiplicityExtPndiv(unsigned long N, vector<unsigned long> primes0)
// this one with externally supplied primes and only counting divisors
{
    if (N == 1) return 1; // corner case
    vector<unsigned long> factors = IncompletePrimeFactor(N, primes0);
    // vector<pair <unsigned long,int>> mfactors;
    // only counting divisors here, turn off all the factor/multiplicity collection 
    unsigned long ndiv = 1;
    if (factors.empty()) // prime
    {
        // mfactors.push_back(make_pair(N,1));
        // return mfactors;
        return 2; 
    }
    vector<unsigned long>::iterator itr;
    int m;
    for (itr = factors.begin(); itr != factors.end(); itr++) {
        m = 0;
        while (N % *itr == 0) {
            m++;
            N /= *itr;
        }
        // mfactors.push_back(make_pair(*itr,m));
        ndiv *= m + 1;    
    }
    if (N != 1) {
        ndiv *= 2; // case of one additional prime factor > sqrt(N)
        // mfactors.push_back(make_pair(N, 1));
    }
    return ndiv;
}

int main() {
    int limdiv = 500;
    int ndiv = 0;
    unsigned long n = 1;
    clock_t c_start = clock();
    vector<unsigned long> primes = Erat(1000);
    while (ndiv < limdiv) {
        if (n % 2 == 0) ndiv = FactorMultiplicityExtPndiv(n/2, primes)*FactorMultiplicityExtPndiv(n+1, primes);
        else ndiv = FactorMultiplicityExtPndiv(n, primes)*FactorMultiplicityExtPndiv((n+1)/2, primes);
        n++;
    }
    n--;
    clock_t c_end = clock();
    cout << endl;
    cout << "n = " << n << " T = " << (n+1)*n/2 << " number of divisors = " << ndiv << endl;
    cout << 1000.0*(c_end - c_start) / CLOCKS_PER_SEC << " ms\n";
    return 0;
}